File: bindings/go/controlplane_test.go
package blazen

// Smoke tests for the control-plane UniFFI surface. These do not require
// a running control-plane server — they exercise the type-construction
// path through the generated bindings and confirm the link table picked
// up the new symbols. End-to-end RPC behaviour is covered by the Rust
// integration suite in blazen-controlplane.

import (
	"strings"
	"testing"

	uniffiblazen "github.com/zachhandley/Blazen/bindings/go/internal/uniffi/blazen"
)

// TestControlPlaneRecordTypesConstruct: the generated record types can
// be constructed by the Go code without panicking. The fields exist as
// expected and the discriminated-union enum is well-formed.
func TestControlPlaneRecordTypesConstruct(t *testing.T) {
	cap := uniffiblazen.ControlPlaneWorkerCapability{
		Kind:    "workflow:hello",
		Version: 1,
	}
	if cap.Kind != "workflow:hello" || cap.Version != 1 {
		t.Fatalf("capability fields lost: %+v", cap)
	}

	mode := uniffiblazen.ControlPlaneAdmissionModeFixed
	if mode != uniffiblazen.ControlPlaneAdmissionModeFixed {
		t.Fatal("admission mode constant mismatch")
	}

	maxInFlight := uint32(4)
	admission := uniffiblazen.ControlPlaneAdmission{
		Mode:        mode,
		MaxInFlight: &maxInFlight,
		TotalMb:     nil,
	}
	if admission.MaxInFlight == nil || *admission.MaxInFlight != 4 {
		t.Fatalf("max_in_flight lost: %+v", admission)
	}

	req := uniffiblazen.ControlPlaneSubmitRequest{
		WorkflowName:    "summarize",
		InputJson:       `{"text":"hi"}`,
		WorkflowVersion: nil,
		RequiredTags:    []string{"region=us-west"},
		IdempotencyKey:  nil,
		DeadlineMs:      nil,
		WaitForWorker:   true,
	}
	if req.WorkflowName != "summarize" || !req.WaitForWorker {
		t.Fatalf("submit request fields lost: %+v", req)
	}

	status := uniffiblazen.ControlPlaneRunStatusPending
	if status != uniffiblazen.ControlPlaneRunStatusPending {
		t.Fatal("run status constant mismatch")
	}
}

// TestControlPlaneClientConnectRejectsBadEndpoint: the blocking
// constructor surfaces transport errors as the typed BlazenException /
// ControlPlaneError variant for malformed URIs. Tests the error path
// of the FFI without needing a server.
func TestControlPlaneClientConnectRejectsBadEndpoint(t *testing.T) {
	ensureInit()
	_, err := uniffiblazen.ControlPlaneClientConnectBlocking("not a uri")
	if err == nil {
		t.Fatal("expected bad endpoint to be rejected")
	}
	msg := err.Error()
	if !strings.Contains(strings.ToLower(msg), "transport") &&
		!strings.Contains(strings.ToLower(msg), "invalid") {
		t.Fatalf("expected transport/invalid error, got %q", msg)
	}
}

// TestControlPlaneWorkerConstructValidatesEndpoint: the blocking worker
// constructor rejects a malformed endpoint URI without needing a server
// to connect to.
func TestControlPlaneWorkerConstructValidatesEndpoint(t *testing.T) {
	ensureInit()
	_, err := uniffiblazen.ControlPlaneWorkerNewBlocking(
		"not a uri",
		"node-test",
		[]uniffiblazen.ControlPlaneWorkerCapability{
			{Kind: "workflow:hello", Version: 1},
		},
	)
	if err == nil {
		t.Fatal("expected bad endpoint URI to be rejected")
	}
	if !strings.Contains(strings.ToLower(err.Error()), "transport") &&
		!strings.Contains(strings.ToLower(err.Error()), "invalid") {
		t.Fatalf("expected transport/invalid error, got %q", err.Error())
	}
}

File: bindings/kotlin/src/test/kotlin/dev/zorpx/blazen/ControlPlaneTest.kt
package dev.zorpx.blazen

import dev.zorpx.blazen.uniffi.BlazenException
import dev.zorpx.blazen.uniffi.ControlPlaneAdmission
import dev.zorpx.blazen.uniffi.ControlPlaneAdmissionMode
import dev.zorpx.blazen.uniffi.ControlPlaneClient
import dev.zorpx.blazen.uniffi.ControlPlaneRunStatus
import dev.zorpx.blazen.uniffi.ControlPlaneSubmitRequest
import dev.zorpx.blazen.uniffi.ControlPlaneWorker
import dev.zorpx.blazen.uniffi.ControlPlaneWorkerCapability
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * Smoke + shape tests for the control-plane UniFFI surface. These do
 * not require a running control-plane server — they exercise the
 * type-construction path and the error path of the FFI without
 * standing up a transport.
 */
class ControlPlaneTest {
    @Test
    fun `worker capability constructs and keeps fields`() {
        val cap = ControlPlaneWorkerCapability(kind = "workflow:hello", version = 1u)
        assertEquals("workflow:hello", cap.kind)
        assertEquals(1u, cap.version)
    }

    @Test
    fun `admission with fixed mode keeps max in flight`() {
        val admission = ControlPlaneAdmission(
            mode = ControlPlaneAdmissionMode.FIXED,
            maxInFlight = 4u,
            totalMb = null,
        )
        assertEquals(ControlPlaneAdmissionMode.FIXED, admission.mode)
        assertEquals(4u, admission.maxInFlight)
        assertEquals(null, admission.totalMb)
    }

    @Test
    fun `admission with reactive mode keeps both nullable fields null`() {
        val admission = ControlPlaneAdmission(
            mode = ControlPlaneAdmissionMode.REACTIVE,
            maxInFlight = null,
            totalMb = null,
        )
        assertEquals(ControlPlaneAdmissionMode.REACTIVE, admission.mode)
        assertEquals(null, admission.maxInFlight)
        assertEquals(null, admission.totalMb)
    }

    @Test
    fun `submit request constructs with all required fields`() {
        val req = ControlPlaneSubmitRequest(
            workflowName = "summarize",
            inputJson = """{"text":"hello"}""",
            workflowVersion = null,
            requiredTags = listOf("region=us-west"),
            idempotencyKey = "dedupe-1",
            deadlineMs = 60_000UL,
            waitForWorker = true,
        )
        assertEquals("summarize", req.workflowName)
        assertTrue(req.waitForWorker)
        assertEquals(listOf("region=us-west"), req.requiredTags)
    }

    @Test
    fun `run status enum has all five variants`() {
        // Sanity-check the wire enum surface — if a new variant was
        // dropped or renamed accidentally, this fails to compile.
        val statuses = listOf(
            ControlPlaneRunStatus.PENDING,
            ControlPlaneRunStatus.RUNNING,
            ControlPlaneRunStatus.COMPLETED,
            ControlPlaneRunStatus.FAILED,
            ControlPlaneRunStatus.CANCELLED,
        )
        assertEquals(5, statuses.size)
    }

    @Test
    fun `client blocking constructor surfaces transport error on bad URI`() {
        val err = assertThrows(BlazenException::class.java) {
            ControlPlaneClient.connectBlocking("not a uri")
        }
        assertNotNull(err)
        val msg = (err.message ?: "").lowercase()
        assertTrue(
            msg.contains("transport") || msg.contains("invalid") || msg.contains("uri"),
            "expected transport/invalid/uri in error message, got '$msg'",
        )
    }

    @Test
    fun `worker blocking constructor validates endpoint URI eagerly`() {
        val err = assertThrows(BlazenException::class.java) {
            ControlPlaneWorker.newBlocking(
                endpoint = "not a uri",
                nodeId = "node-test",
                capabilities = listOf(
                    ControlPlaneWorkerCapability("workflow:hello", 1u),
                ),
            )
        }
        assertNotNull(err)
        val msg = (err.message ?: "").lowercase()
        assertTrue(
            msg.contains("transport") || msg.contains("invalid") || msg.contains("uri"),
            "expected transport/invalid/uri in error message, got '$msg'",
        )
    }
}
